### DISCUSSION TITLE
NanoMQ clustering using kubernetes, ECS, or just DNS load balancing

### DISCUSSION THREAD

**Discussion by brentpicasso** [nanomq/nanomq#1233] [Q&A] [answered]
Hi, we are extremely interested to use NanoMQ, especially with the ZMQ gateway option. 

For horizontal scaling, we were thinking of using some kind of load balancing technique using kubernetics, ECS, or just simple DNS load balancing across a cluster of computers. 

Any reason something like this would not work?  The clustering needed would need to support persistent connections, naturally.
https://www.intuz.com/blog/how-to-deploy-mqtt-broker-in-a-kubernetes-cluster

Thanks in advance!

**Comment by JaylinYu** [comment]
Clustering is certainly on my roadmap, but it is not sth gonna happen overnight...
Using mosquitto/nanomq with Kubernetes in cluster mode doesn't solve the distributed consistency. The clients from different node cannot talks to each other seamlessly like they are connected in single instance

**Reply by brentpicasso** [reply]
Thanks for the reply. 

If there was a cluster of nanoMQ servers, with some kind of load balancing scheme, how well would it work right now? I suppose we would need to solve our own distributed consistency problem, correct? 

For example, with DNS load balancing, a client would establish a persistent connection to a node, and if the connection dropped, a client could re-establish a connection with the same or a different node.

**Reply by JaylinYu** [reply]
Sure, you can manage multiple instances of NanoMQ with DNS load balancing. It could work, but clients connects to different node will not able to talk with each other, Idk if this is also within your expectation.

**Reply by brentpicasso** [reply]
Thanks for the reply. 

Yes, we will need to handle the concurrency with an external agent that probably runs on the same host. 

IoT device ---> NanoMQ --(zeromq agent)----> stream receiver agent --->  database / rest of system

These agents will be responsible for the real-time processing of the data (writing to the filesystem, database) and must be able to handle the case when an iOT device reconnects to a different NanoMQ instance, avoiding race conditions, etc.

**Reply by JaylinYu** [reply]
Yup, What I suggest for now is:
Setting up a routing model that defines the corresponding relationship between NanoMQ nodes and clients according to messaging requirements, like A1 needs to talk with A2-A10, then they should be connected to the same NanoMQ node, and direct A1-10 clients to target node via LB.

**Reply by JaylinYu** [reply]
Your paradigm looks fine to me, 
 |________ /   NanoMQ \
IoT device --- NanoMQ -  agent (knows what client in which nodes) --- ......
_________ \    NanoMQ /

**Reply by brentpicasso** [reply, marked as answer]
Thanks for the replies. 

Right now, the architecture does not require IoT nodes to talk to each other - it is primarily about the transmission of telemetry data from many IoT devices into storage. 

So, in a scenario where an IoT device connects to one node, it will broadcast data and be stored in a database/filesystem. 

If the IoT device disconnects and re-connects to a different node, we need to ensure the data persistence continues seamlessly.

**Comment by JaylinYu** [comment]
BTW, I would like to implement basic clustering at the end of the this year.

**Reply by chm90** [reply]
How is it going?

**Reply by JaylinYu** [reply]
I am focusing more on SDV now. This is not on my plan anymore

**Reply by brentpicasso** [reply]
What's a SDV?
